text/0000-confidential-token-ids.md

```diff
@@ -0,0 +1,270 @@
+
+Untrusted proceeds to [sort the WellFormedTxContext](https://github.com/mobilecoinfoundation/mobilecoin/blob/a092039a4a5cc7f5e3e16eeadd0b2bc3a12667ae/consensus/enclave/api/src/lib.rs#L131) by `fee_ratio` as part of the [nominate phase](https://github.com/mobilecoinfoundation/mobilecoin/blob/cfa51d26ae943a9055698bb209c2fe06fa7a7cac/consensus/scp/src/slot.rs#L613).
+
+Now that the Tx is validated as "well-formed," sorted and combined for a ballot with other transactions, and is a candidate for inclusion in the consensus slot, the WellFormedEncryptedTx is [broadcast](https://github.com/mobilecoinfoundation/mobilecoin/blob/2f90154a445c769594dfad881463a2d4a003d7d6/peers/src/traits.rs#L28) to the node's attested peers as required by the consensus protocol, using the [`peer_tx_propose`](https://github.com/mobilecoinfoundation/mobilecoin/blob/2d5190e60f6820cebcd585c19d16cecf9ba4d89b/consensus/enclave/impl/src/lib.rs#L328) attested endpoint.
+
+The TxManager in untrusted for each node maintains a cache of encrypted txs and tx contexts that its enclave has found to be well-formed. Once the consensus protocol is complete, the untrusted TxManager calls [form_block]() which prompts the [enclave to `form_block`](https://github.com/mobilecoinfoundation/mobilecoin/blob/cfa51d26ae943a9055698bb209c2fe06fa7a7cac/consensus/enclave/impl/src/lib.rs#L421) using the full transaction contents.
+
+Because untrusted never sees the `fee` or `token_id` in the clear, the attacker would need to sustain an enclave exploit, such as a side channel attack, that would allow them to access data while the enclave is processing to discern the TokenType of a transaction. With this information, the attacker could, for the duration of the exploit, determine which asset type each transaction was composed of, and could also construct a transaction graph for statistical analysis of the inputs, which are also only visible in the TxPrefix, but not persisted to the chain. Once the integrity of SGX was restored, the attacker would no longer be able to obtain information about the transaction inputs or TokenType, and forward secrecy is preserved. Further, by being thoughtful about the memory access of the `token_id`, and using constant time operations, we can mitigate the data leakage even in the event of a side channel exploit.
+
+### Upgrade Plan
+
+As this is a substantial change to the ledger format and transaction protocol, we need to be thoughtful about what rollout looks like for this change to land with minimal impact on existing clients and maximum backward compatibility.
+
+The rollout proposal is the following:
```

Comment:
I want to suggest that we make a second MCIP for the roll out proposal -- I actually think we may need to rethink the plan for rolling out recoverable transaction history to use a similar mechanism as this

Reply:
https://github.com/mobilecoinfoundation/mcips/pull/26